util/converter/tarconverter/tarconverter.go

```diff
@@ -0,0 +1,82 @@
+package tarconverter
+
+import (
+	"archive/tar"
+	"errors"
+	"io"
+)
+
+type HeaderConverter func(*tar.Header)
+
+func NewWriter(w io.Writer, f HeaderConverter) io.WriteCloser {
+	pr, pw := io.Pipe()
+	tr := tar.NewReader(pr)
+	tw := tar.NewWriter(w)
+	x := &writer{
+		pr: pr,
+		pw: pw,
+		tr: tr,
+		tw: tw,
+		ch: make(chan struct{}),
+	}
+	go x.routine(f)
+	return x
+}
+
+type writer struct {
+	pr *io.PipeReader
+	pw *io.PipeWriter
+	tr *tar.Reader
+	tw *tar.Writer
+	ch chan struct{}
+}
+
+func (x *writer) Write(p []byte) (int, error) {
+	return x.pw.Write(p)
+}
+
+func (x *writer) Close() error {
+	_ = x.pw.Close()
+	<-x.ch
+	return nil
+}
+
+func (x *writer) routine(f HeaderConverter) {
+	var err error
+	for {
+		var hdr *tar.Header
+		hdr, err = x.tr.Next()
+		if err != nil {
+			break
+		}
+		if hdr == nil {
+			err = errors.New("got nil header")
+			break
+		}
+		f(hdr)
+		if err = x.tw.WriteHeader(hdr); err != nil {
+			break
+		}
+		if _, err = io.Copy(x.tw, x.tr); err != nil {
+			break
+		}
+	}
+	// usually err is io.EOF here
+	_ = x.tw.Close()
+	_ = x.pr.CloseWithError(err)
+	close(x.ch)
+}
+
+func NewReader(r io.Reader, f HeaderConverter) io.ReadCloser {
+	pr, pw := io.Pipe()
+	w := NewWriter(pw, f)
```

Comment:
Looks like we are using 2 `io.Pipe` in here as there is another one inside the `NewWriter`. I think only one should be needed to handle this case.

Reply:
Can be a follow-up?

Reply:
Thanks, fixed